agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -197,52 +191,81 @@ public Function thenSort(@NonNull final Comparator comparator) {
     @SuppressWarnings("unchecked")
     public Object apply(@NonNull final Object input) {
       Object item = input;
-      for (final Function function : functions) {
-        item = function.apply(item);
-      }
-      return item;
-    }
-  }
-
-  private static final class FilterFunction<T> implements Function<List<T>, List<T>> {
-    @NonNull
-    private final Predicate filter;
-
-    FilterFunction(@NonNull final Predicate filter) {
-      this.filter = checkNotNull(filter);
-    }
+      boolean compilerList = false;
+      int i = 0;
+      final int length = functions.length;
+      while (i < length) {
+        final int type = (int) functions[i++];
+        final Object command = functions[i++];
+        switch (type) {
+          case LIMIT: {
+            final List list = (List) item;
+            final int limit = (int) command;
+            if (limit <= 0) {
+              item = emptyList();
+            } else if (list.size() < limit) {
+              item = input;
+            } else {
+              final List subList = list.subList(0, limit);
+              item = compilerList ? subList : new ArrayList<>(subList);
+              compilerList = true;
+            }
+          }
+          break;
+          case FILTER: {
+            final List list = (List) item;
+            if (list.isEmpty()) {
+              item = emptyList();
+            } else {
+              final int size = list.size();
+              final List result = new ArrayList(size);
+              for (int j = 0; j < size; j++) {
+                final Object listItem = list.get(j);
+                if (((Predicate) command).apply(listItem)) {
+                  result.add(listItem);
+                }
+              }
+              item = result;
+              compilerList = true;
+            }
+          }
+          break;
+          case SORT: {
+            final List output = compilerList ? ((List) item) : new ArrayList<>(((List) item));
+            Collections.sort(output, (Comparator) command);
+            item = output;
+            compilerList = true;
+          }
+          break;
+          case MAP: {
+            final List list = (List) item;
+            final int size = list.size();
+            if (list.isEmpty()) {
+              item = emptyList();
+            } else if (compilerList) {
+              for (int j = 0; j < size; j++) {
+                list.set(j, ((Function) command).apply(list.get(j)));
```

Comment:
this cast probably incurs runtime type-checking overhead, so do it outside the for-loop.